src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+                values->m_precomputed.m_sigma_a =
+                sigma_a_from_melanin(values->m_melanin, values->m_melanin_redness);
+            else
+                values->m_precomputed.m_sigma_a =
+                sigma_a_from_rgb(values->m_reflectance, values->m_beta_N);
+
+            // Longitudinal roughness parameter
+            values->m_precomputed.m_v =
+                0.726f * values->m_beta_M +
+                0.812f * values->m_beta_M * values->m_beta_M +
+                3.7f * pow_int(values->m_beta_M, 20);
+            values->m_precomputed.m_v *= values->m_precomputed.m_v;
+
+            // Scale factor to convert logistic function to gaussian
+            values->m_precomputed.m_s =
+                0.626657069f * (0.265f * values->m_beta_N +
+                1.194f * values->m_beta_N * values->m_beta_N +
+                5.372f * pow_int(values->m_beta_N, 22));
+
+        }
+
+        void sample(
+            SamplingContext&            sampling_context,
+            const void*                 data,
+            const bool                  adjoint,
+            const bool                  cosine_mult,
+            const int                   modes,
+            BSDFSample&                 sample) const override
+        {
+            if (!ScatteringMode::has_glossy(modes))
+                return;
+
+            const InputValues* values = static_cast<const InputValues*>(data);
+            Vector3f outgoing = sample.m_outgoing.get_value();
+            const Vector3f wo = sample.m_shading_basis.transform_to_local(outgoing);
+
+            sampling_context.split_in_place(2, 1);
+            Vector2f sample_M = sampling_context.next2<Vector2f>();
+            sampling_context.split_in_place(2, 1);
+            Vector2f sample_N = sampling_context.next2<Vector2f>();
+
+            float sin_theta_o = wo.x;
+            float cos_theta_o = std::sqrt(std::max(0.0f, 1.0f - sin_theta_o * sin_theta_o));
+            float phi_o = std::atan2(wo.y, wo.z);
+
+            float sin_theta_t = sin_theta_o / values->m_eta;
+            float cos_theta_t = std::sqrt(std::max(0.0f, 1.0f - sin_theta_t * sin_theta_t));
+
+            float eta_p = std::sqrt(values->m_eta * values->m_eta - sin_theta_o * sin_theta_o) / cos_theta_o;
+            float sin_gamma_t = values->m_precomputed.m_h / eta_p;
+            float cos_gamma_t = std::sqrt(std::max(0.0f, 1.0f - sin_gamma_t * sin_gamma_t));
+            float gamma_t = std::asin(clamp(sin_gamma_t, -1.0f, 1.0f));
+            float gamma_o = std::asin(clamp(values->m_precomputed.m_h, -1.0f, 1.0f));
+
+            // Compute attenuation for the single path through the hair cylinder
+            Spectrum T = exp(-values->m_precomputed.m_sigma_a * (( 2.f * cos_gamma_t) / cos_theta_t));
+
+            // Determine which scattering component to sample for hair scattering
+            std::array<Spectrum, 4> ap = attenuation(cos_theta_o, values->m_eta, values->m_precomputed.m_h, T);
```

Comment:
I'm not sure returning an array of `Spectrum` is going to be optimized properly. I'd rather we pass it by (mutable) ref and let `attenuation()` fill it.